### PR TITLE
Minor: improve GroupsAccumulator and Accumulator documentation

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of #7013 

## Rationale for this change

We recently made `GroupsAccumulator` part of the public API in https://github.com/apache/arrow-datafusion/pull/8892 (thank you @guojidan ❤️ ). Now that it is public it would be nice to have some additional user facing documentation / rationale

## What changes are included in this PR?
Update docstrings with explanation of the difference between `Accumulator` and `GroupsAccumulator`

## Are these changes tested?
Existing tests

## Are there any user-facing changes?
Doc only

### CODE REVIEW

File: datafusion/expr/src/accumulator.rs [hunk 1]
```diff
@@ -21,11 +21,13 @@ use arrow::array::ArrayRef;
 use datafusion_common::{internal_err, DataFusionError, Result, ScalarValue};
 use std::fmt::Debug;
 
-/// Describes an aggregate functions's state.
+/// Tracks an aggregate functions's state.
```
Comment (Jefffrey) [inline datafusion/expr/src/accumulator.rs:24]:
```suggestion
/// Tracks an aggregate function's state.
```

File: datafusion/expr/src/accumulator.rs [hunk 2]
```diff
@@ -21,11 +21,13 @@ use arrow::array::ArrayRef;
...
 ///
-/// `Accumulator`s are stateful objects that live throughout the
-/// evaluation of multiple rows and aggregate multiple values together
-/// into a final output aggregate.
+/// `Accumulator`s are stateful objects that implement a single group. They
+/// aggregate values from multiple rows together into a final output aggregate.
+///
+/// [`GroupsAccumulator]` is a additional more performant (but also complex) API
```
Comment (Jefffrey) [inline datafusion/expr/src/accumulator.rs:29]:
```suggestion
/// [`GroupsAccumulator]` is an additional more performant (but also complex) API
```

File: datafusion/expr/src/groups_accumulator.rs [hunk 3]
```diff
@@ -59,12 +59,24 @@ impl EmitTo {
...
+/// optional and is harder to implement than `Accumulator`, but can be much
+/// faster for queries with many group values.  See the [Aggregating Millions of
+/// Groups Fast blog] for more background.
+///
+/// # Details
 /// Each group is assigned a `group_index` by the hash table and each
-/// accumulator manages the specific state, one per group_index.
+/// accumulator manages the specific state, one per `group_index`.
 ///
 /// group_indexes are contiguous (there aren't gaps), and thus it is
```
Comment (Jefffrey) [inline datafusion/expr/src/groups_accumulator.rs:74]:
```suggestion
/// `group_index`es are contiguous (there aren't gaps), and thus it is
```

**Review by Jefffrey — APPROVED** [review]
Looks good :+1:

**Comment by alamb** [comment]
Thanks again for the review @Jefffrey

**Pull request merged**